retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java

```diff
@@ -0,0 +1,179 @@
+    public MeterIdPrefix apply(MeterRegistry registry, RequestLog log) {
+        final ImmutableList.Builder<Tag> tagListBuilder = ImmutableList.builder();
+        buildTags(tagListBuilder, log);
+        RequestMetricSupport.appendHttpStatusTag(tagListBuilder, log);
+        return new MeterIdPrefix(name, tagListBuilder.build());
+    }
+
+    private void buildTags(ImmutableList.Builder<Tag> tagsBuilder, RequestLog log) {
+        final Invocation invocation = InvocationUtil.getInvocation(log);
+
+        tagsBuilder.add(Tag.of(serviceTagName, serviceName));
+        if (invocation != null) {
+            final String methodName = invocation.method().getName();
+            final List<Tag> methodTags = methodNameToTags.get(methodName);
+            if (methodTags != null) {
+                tagsBuilder.addAll(methodTags);
+                return;
+            }
+        }
+
+        tagsBuilder.add(Tag.of("method", log.method().name()));
+    }
+
+    @VisibleForTesting
+    static Map<String, List<Tag>> defineTagsForMethods(Class serviceClass) {
+        final Builder<String, List<Tag>> methodNameToTags = ImmutableMap.builder();
+
+        final Method[] declaredMethods = serviceClass.getDeclaredMethods();
+        for (final Method clientServiceMethod : declaredMethods) {
+            final Annotation[] declaredAnnotations = clientServiceMethod.getDeclaredAnnotations();
+            if (declaredAnnotations.length == 0) {
+                continue;
+            }
+
+            for (final Annotation annotation : declaredAnnotations) {
+                if (!retrofitAnnotations.contains(annotation.annotationType())) {
+                    continue;
+                }
+
+                final ImmutableList.Builder<Tag> tags = ImmutableList.builder();
+                tags.addAll(tagsFromAnnotation(annotation));
+                tags.add(Tag.of("method", clientServiceMethod.getName()));
+                methodNameToTags.put(clientServiceMethod.getName(), tags.build());
+            }
+        }
+
+        return methodNameToTags.build();
+    }
+
+    private static List<Tag> tagsFromAnnotation(Annotation annotation) {
+        final String httpMethod;
+        final String httpPath;
+        if (annotation.annotationType().equals(HTTP.class)) {
+            final HTTP http = (HTTP) annotation;
+            httpMethod = http.method();
+            httpPath = http.path();
+        } else {
+            final Method valueMethod;
+            try {
+                valueMethod = annotation.getClass().getMethod("value");
+                httpPath = (String) valueMethod.invoke(annotation);
+            } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+                // Should never happen on valid Retrofit client.
+                throw new IllegalStateException("Provided Retrofit client have unexpected annotation: " +
+                                                annotation.annotationType(), e);
```

Comment:
Isn't it better skipping rather than raising an exception? A user may have added non-Retrofit annotations for other purposes.

Reply:
Ah I see. Thanks for explanation.